hibernate-core/src/main/java/org/hibernate/event/internal/DefaultMergeEventListener.java

```diff
@@ -319,17 +382,25 @@ private void saveTransientEntity(
 		}
 	}
 
-	protected void entityIsDetached(MergeEvent event, MergeContext copyCache) {
+	protected void entityIsDetached(MergeEvent event, Object copiedId, Object originalId, MergeContext copyCache) {
 		LOG.trace( "Merging detached instance" );
 
 		final Object entity = event.getEntity();
 		final EventSource source = event.getSession();
 		final EntityPersister persister = source.getEntityPersister( event.getEntityName(), entity );
 		final String entityName = persister.getEntityName();
-
-		Object id = getDetachedEntityId( event, entity, persister );
+		if ( originalId == null ) {
+			originalId = persister.getIdentifier( entity, source );
+		}
+		final Object clonedIdentifier;
+		if ( copiedId == null ) {
+			clonedIdentifier = persister.getIdentifierType().deepCopy( copiedId, source.getFactory() );
```

Comment:
Looks like tests are not covering this case yet because you try to copy `copiedId` which is always null here.

```suggestion
			clonedIdentifier = persister.getIdentifierType().deepCopy( originalId, source.getFactory() );
```